builder/actions.py

```diff
@@ -51,10 +51,37 @@ def delete_search(*, search):
     # Grab the PK before we delete the instance
     search_pk = search.pk
 
-    # Delete any codes that only belong to this search
-    search.version.code_objs.annotate(num_results=Count("results")).filter(
-        results__search=search, num_results=1
-    ).delete()
+    # Delete any codes that:
+    # - only belong to this search
+    # - are not included
+    # - are not descendants of an included code
+
+    # Find all code objs that belong to this search and no others
+    search_only_code_objs = search.version.code_objs.annotate(
+        num_results=Count("results")
+    ).filter(results__search=search, num_results=1)
+    # Get all explicitly included codes on the codelist
+    all_included_codes = search.version.codeset.codes()
+
+    # Build a set of codes_to_keep from this search, consisting of any included codes, their descendants,
+    # and any codes with ancestor that is included
+    hierarchy = search.version.hierarchy
+    codes_to_keep = set()
+
+    for code_obj in search_only_code_objs:
+        # Keep a code if it is explicitly included or a descendant of an explicitly included code
+        if code_obj.code in all_included_codes:
```

Comment:
I'm fairly sure we don't need this branch, and that the following will do what we need:

```
    codes_to_keep = []
    for code_obj in search_only_code_objs:
        ancestors = hierarchy.ancestors(code_obj.code)
        ancestors_in_included = ancestors & all_included_codes
        if ancestors_in_included:
            codes_to_keep.append(code_obj.code)
```

Reply:
Almost - it needed to be:
```
    codes_to_keep = []
    for code_obj in search_only_code_objs:
        ancestors = hierarchy.ancestors(code_obj.code)
        ancestors_in_included = ancestors & all_included_codes
        if ancestors_in_included or code_obj.code in all_included_codes:
            codes_to_keep.append(code_obj.code)
```
I've updated this, and also added another test since the previous one didn't catch that extra `or code_obj.code in all_included_codes` that was needed.